src/client.rs

```diff
@@ -280,3 +270,368 @@ where
+                }
+                TransactionProgressStatus::Invalid => {
+                    return Err(TransactionError::Invalid.into())
+                }
+                TransactionProgressStatus::Usurped(_) => {
+                    return Err(TransactionError::Usurped.into())
+                }
+                TransactionProgressStatus::Dropped => {
+                    return Err(TransactionError::Dropped.into())
+                }
+                // Ignore and wait for next status event:
+                _ => continue,
+            }
+        }
+        Err(RpcError::Custom("RPC subscription dropped".into()).into())
+    }
+
+    /// Wait for the transaction to be finalized, and return a [`TransactionInBlock`]
+    /// instance when it is, or an error if there was a problem waiting for finalization.
+    ///
+    /// **Note:** consumes `self`. If you'd like to perform multiple actions as the state of the
+    /// transaction progresses, use [`TransactionProgress::next()`] instead.
+    pub async fn wait_for_finalized(
+        mut self,
+    ) -> Result<TransactionInBlock<'client, T>, Error> {
+        while let Some(status) = self.next().await? {
+            match status {
+                // Finalized! Return.
+                TransactionProgressStatus::Finalized(s) => return Ok(s),
+                // Error scenarios; return the error.
+                TransactionProgressStatus::FinalityTimeout(_) => {
+                    return Err(TransactionError::FinalitySubscriptionTimeout.into())
+                }
+                TransactionProgressStatus::Invalid => {
+                    return Err(TransactionError::Invalid.into())
+                }
+                TransactionProgressStatus::Usurped(_) => {
+                    return Err(TransactionError::Usurped.into())
+                }
+                TransactionProgressStatus::Dropped => {
+                    return Err(TransactionError::Dropped.into())
+                }
+                // Ignore and wait for next status event:
+                _ => continue,
+            }
+        }
+        Err(RpcError::Custom("RPC subscription dropped".into()).into())
+    }
+
+    /// Wait for the transaction to be finalized, and for the transaction events to indicate
+    /// that the transaction was successful. Returns the events associated with the transaction,
+    /// as well as a couple of other details (block hash and extrinsic hash).
+    ///
+    /// **Note:** consumes self. If you'd like to perform multiple actions as progress is made,
+    /// use [`TransactionProgress::next()`] instead.
+    pub async fn wait_for_finalized_success(self) -> Result<TransactionEvents<T>, Error> {
+        let evs = self.wait_for_finalized().await?.wait_for_success().await?;
+        Ok(evs)
+    }
+}
+
+/// Possible transaction progess states returned from our [`TransactionProgress::next()`] call.
+#[derive(Debug)]
+pub enum TransactionProgressStatus<'client, T: Config> {
```

Comment:
Could possibly be shortened into `TransactionStatus`?

Reply:
Oooh, I wanted to keep it separate from `TransactionStatus` which also exists in this crate, although I realise now that that enum isn't actually exposed to the public anywhere, so I'll make this name nicer and make that one uglier.